applications/plugins/SofaCUDA/sofa/gpu/cuda/CudaTriangularFEMForceFieldOptim.cu

```diff
@@ -45,207 +38,192 @@ struct GPUTriangleInfo
-        const TriangleInfo& ti = triInfo[i];
-        TriangleState& ts = triState[i];
-        CudaVec3<float> a  = x[t.ia];
-        CudaVec3<float> ab = x[t.ib]-a;
-        CudaVec3<float> ac = x[t.ic]-a;
-        //computeTriangleRotation(ts.frame, ab, ac);
-        CudaVec3<float> frame_x = ab;
-        CudaVec3<float> n = cross(ab,ac);
-        CudaVec3<float> frame_y = cross(n,ab);
-        frame_x *= invnorm(frame_x);
-        frame_y *= invnorm(frame_y);
-        ts.frame_x = frame_x;
-        ts.frame_y = frame_y;
-
-        float dbx = ti.bx - frame_x*ab;
-        // float dby = 0
-        float dcx = ti.cx - frame_x*ac;
-        float dcy = ti.cy - frame_y*ac;
-        //sout << "Elem" << i << ": D= 0 0  " << dbx << " 0  " << dcx << " " << dcy << sendl;
-
-        CudaVec3<float> strain = CudaVec3<float>::make (
-            ti.cy * dbx,                // ( cy,   0,  0,  0) * (dbx, dby, dcx, dcy)
-            ti.bx * dcy,                // (  0, -cx,  0, bx) * (dbx, dby, dcx, dcy)
-            ti.bx * dcx - ti.cx * dbx); // (-cx,  cy, bx,  0) * (dbx, dby, dcx, dcy)
-
-        float gammaXY = (strain.x+strain.y)*gamma;
-
-        CudaVec3<float> stress = CudaVec3<float>::make (
-            mu*strain.x + gammaXY,    // (gamma+mu, gamma   ,    0) * strain
-            mu*strain.y + gammaXY,    // (gamma   , gamma+mu,    0) * strain
-            (float)(0.5)*mu*strain.z); // (       0,        0, mu/2) * strain
-
-        ts.stress = stress;
-
-        stress *= ti.ss_factor;
-        //sout << "Elem" << i << ": F= " << -(ti.cy * stress[0] - ti.cx * stress[2] + ti.bx * stress[2]) << " " << -(ti.cy * stress[2] - ti.cx * stress[1] + ti.bx * stress[1]) << "  " << (ti.cy * stress[0] - ti.cx * stress[2]) << " " << (ti.cy * stress[2] - ti.cx * stress[1]) << "  " << (ti.bx * stress[2]) << " " << (ti.bx * stress[1]) << sendl;
-        CudaVec3<float> fb = frame_x * (ti.cy * stress.x - ti.cx * stress.z)  // (cy,   0, -cx) * stress
-                + frame_y * (ti.cy * stress.z - ti.cx * stress.y); // ( 0, -cx,  cy) * stress
-        CudaVec3<float> fc = frame_x * (ti.bx * stress.z)                      // ( 0,   0,  bx) * stress
-                + frame_y * (ti.bx * stress.y);                     // ( 0,  bx,   0) * stress
-        CudaVec3<float> fa = -fb-fc;
-
-#if defined(__CUDA_ARCH__) &&  __CUDA_ARCH__ < 200
-        f[t.ia] += fa;
-        f[t.ib] += fb;
-        f[t.ic] += fc;
-#else
-        atomicAdd(&(f[t.ia].x), fa.x);
-        atomicAdd(&(f[t.ia].y), fa.y);
-        atomicAdd(&(f[t.ia].z), fa.z);
-        atomicAdd(&(f[t.ib].x), fb.x);
-        atomicAdd(&(f[t.ib].y), fb.y);
-        atomicAdd(&(f[t.ib].z), fb.z);
-        atomicAdd(&(f[t.ic].x), fc.x);
-        atomicAdd(&(f[t.ic].y), fc.y);
-        atomicAdd(&(f[t.ic].z), fc.z);
-#endif
+    GPUTriangleInfo t = gpuTriangleInfo[i];
+    const TriangleInfo<real>& ti = triInfo[i];
+    TriangleState<real>& ts = triState[i];
+    
+    CudaVec3 a  = x[t.ia];
```

Comment:
I understand you did not touch this code, but if you have the time, you can constify some of the variables.